### PR TITLE
Trie error handling

### PR DESCRIPTION
**Pull request by zsfelfoldi**

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Nov 25 17:30:56 UTC 2015_

File: trie/trie.go [hunk 1]
```diff
@@ -79,13 +82,16 @@ type Trie struct {
 // New will panics if db is nil or root does not exist in the
```
Comment (zelig) [inline trie/trie.go]: comment does not reflect what it actally does

Comment (zelig) [inline trie/trie.go]: comment does not reflect what it actally does

File: trie/secure_trie.go [hunk 2]
```diff
@@ -73,14 +83,39 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
 // Delete removes any existing value for key from the trie.
 func (t *SecureTrie) Delete(key []byte) {
-	t.Trie.Delete(t.hashKey(key))
+	err := t.TryDelete(key)
+	if err != nil {
+		unhandledError(err)
+	}
+}
+
+// Delete removes any existing value for key from the trie.
```
Comment (zelig) [inline trie/secure_trie.go]: extend comment to differentiate from Delete

File: trie/trie.go [hunk 3]
```diff
@@ -287,34 +380,46 @@ func concat(s1 []byte, s2 ...byte) []byte {
...
+	KeyPrefix, KeySuffix []byte
+}
+
+func (err *MissingNodeError) Error() string {
+	return fmt.Sprintf("Missing trie node %064x", err.NodeHash)
+}
+
+func unhandledError(err error) {
+	if glog.V(logger.Error) {
+		glog.Errorf("Unhandled trie error: %v", err)
```
Comment (karalabe) [inline trie/trie.go]: I don't think it's a good idea to put a log statement in a helper function. The reason is that logs are tagged with line numbers that should provide a hint as to where the issue went wrong. With this approach all error logs will have the same line number, loosing part of its information content.

Further, you can simplify the error reporting to:

```
glog.V(logger.Error).Errorf("Unhandled trie error: %v", err)
```

File: trie/secure_trie.go [hunk 4]
```diff
@@ -73,14 +83,39 @@ func (t *SecureTrie) Get(key []byte) []byte {
 // The value bytes must not be modified by the caller while they are
 // stored in the trie.
 func (t *SecureTrie) Update(key, value []byte) {
+	err := t.TryUpdate(key, value)
+	if err != nil {
+		unhandledError(err)
+	}
```
Comment (karalabe) [inline trie/secure_trie.go]: You can shorten this to `if err := ...; err != nil { ... }`

File: trie/secure_trie.go [hunk 5]
```diff
@@ -73,14 +83,39 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
+// Update associates key with value in the trie. Subsequent calls to
+// Get will return value. If value has length zero, any existing value
+// is deleted from the trie and calls to Get will return nil.
+//
+// The value bytes must not be modified by the caller while they are
+// stored in the trie.
+func (t *SecureTrie) TryUpdate(key, value []byte) error {
 	hk := t.hashKey(key)
-	t.Trie.Update(hk, value)
+	err := t.Trie.TryUpdate(hk, value)
```
Comment (karalabe) [inline trie/secure_trie.go]: Same here, please shorten err assignment and check into the same if construct.

File: trie/secure_trie.go [hunk 6]
```diff
@@ -73,14 +83,39 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
+		return err
+	}
 	t.Trie.db.Put(t.secKey(hk), key)
+	return nil
 }
 
 // Delete removes any existing value for key from the trie.
 func (t *SecureTrie) Delete(key []byte) {
-	t.Trie.Delete(t.hashKey(key))
+	err := t.TryDelete(key)
```
Comment (karalabe) [inline trie/secure_trie.go]: Same, please shorten.

File: trie/trie.go [hunk 7]
```diff
@@ -131,17 +153,41 @@ func (t *Trie) Get(key []byte) []byte {
 // The value bytes must not be modified by the caller while they are
 // stored in the trie.
 func (t *Trie) Update(key, value []byte) {
+	err := t.TryUpdate(key, value)
```
Comment (karalabe) [inline trie/trie.go]: Please shorten.

File: trie/trie.go [hunk 8]
```diff
@@ -185,28 +250,45 @@ func (t *Trie) insert(n node, key []byte, value node) node {
 
 // Delete removes any existing value for key from the trie.
 func (t *Trie) Delete(key []byte) {
+	err := t.TryDelete(key)
```
Comment (karalabe) [inline trie/trie.go]: Please shorten.

File: trie/trie.go [hunk 9]
```diff
@@ -287,34 +380,46 @@ func concat(s1 []byte, s2 ...byte) []byte {
...
-		return t.resolveHash(n)
+		return t.resolveHash(n, prefix, suffix)
+	}
+	return n, nil
+}
+
+type MissingNodeError struct {
+	RootHash, NodeHash   common.Hash
+	KeyPrefix, KeySuffix []byte
+}
```
Comment (karalabe) [inline trie/trie.go]: Any particular reason this is in the middle of the file? :P 

Please either move this to the top of the file, or better yet I would create an errors.go file to contain this. Usually that is the convention throughout our code base.

File: trie/trie.go (hunk 3 above)
Comment (fjl) [inline trie/trie.go]: `Errorf` is not a method, unfortunately.
But you can combine the error and logging checks:

```go
if err := operation(); err != nil && glog.V(logger.Error) {
     glog.Errorf("Unhandled trie error: %v", err)
}
return nil
```

File: trie/proof.go [hunk 10]
```diff
@@ -39,7 +41,11 @@ func (t *Trie) Prove(key []byte) []rlp.RawValue {
 		case nil:
 			return nil
 		case hashNode:
-			tn = t.resolveHash(n)
+			var err error
+			tn, err = t.resolveHash(n, nil, nil)
+			if err != nil && glog.V(logger.Error) {
+				glog.Errorf("Unhandled trie error: %v", err)
```
Comment (fjl) [inline trie/proof.go]: Maybe `return nil` here? The proof will not be useful if the trie is incomplete.

**Comment by fjl** [comment]
:+1:, please squash.

**Comment by zsfelfoldi** [comment]
squashed.

**Comment by zsfelfoldi** [comment]
fixed, added description for trie.MissingNodeError

**Comment by obscuren** [comment]
Minor issue, please format commit messages like `<changes packages>: commit message`, e.g.

```
trie: changed error handling

Changed error handling such that X now performs in such a way that X can handle Z properly.
```

**Comment by zsfelfoldi** [comment]
@obscuren ok, I changed the commit message

**Comment by obscuren** [comment]
:+1:
